design/layer2-bind-interfaces.md

```diff
@@ -0,0 +1,56 @@
+# Configuring announcing interfaces
+
+## Summary
+
+The purpose of this enhancement is to make Speakers only announce VIPs on specified network interfaces.
+
+Previous discussions about this topic can be found in [this issue](https://github.com/metallb/metallb/issues/277).
+
+## Motivation
+
+A system may have multiple networks based on functions, and thus people want MetalLB only works on specified networks.
+In BGP mode, this situation can be handled by BGP configuration, while in L2 mode,  there is no way to achieve this purpose.
+
+[<img src="./images/example_system.png" width="800"/>](./images/example_system.png)
+
+### Goals
+- The network interfaces the MetalLB uses to broadcast VIP can be specified.
+
+### Non-Goals
+- Change any core functions of MetalLB
+
+## Data model
+
+Based on the current MetalLB CRD design (`8d906f0abef3d7ed5b0435fa4edfbe5e37b38557`), we'd like to add a new field `BindInterfaces` to `L2Advertisement`.
+
+`BindInterfaces` describes network interface names of the Nodes MetalLB should use. It is a list of interfaces' names, and it is optional. Once users set "bindInterface", then Speakers should only create announcers on these interfaces. Otherwise, Speakers keep the default behavior which creates announcers on every interface.
+
+The new definition of `L2Advertisement` is:
+```go
+type L2Advertisement struct {
+  Name                     string `yaml:"name"`
+  IPAddressPools           []string `yaml:"ipAddressPools"`
+  IPAddressPoolSelector    *metav1.LabelSelector `yaml:"ipAddressPoolSelector"`
+  BindInterfaces           []string `yaml:"bindInterfaces"`
+}
+```
+
+An example definition can be:
+```yaml
+apiVersion: metallb.io/v1beta1
+kind: L2Advertisement
+metadata:
+  name: example-advertisement
+  namespace: metallb-system
+spec:
+    ipaddresspools:
+      - pool1
+      - pool2
+    bindInterfaces:
+      - eth0
+      - eth1
+```
```

Comment:
Can you extend the example to see if we are able to support the more complex scenario described in https://github.com/metallb/metallb/issues/277#issuecomment-559599874 ? 
I think we'd be (more or less for free), given the L2Advertisement node selector.

Reply:
Sorry, what does the second line mean？Can you describe it more detailly?

Reply:
If you look at https://github.com/metallb/metallb/issues/277#issuecomment-559599874, Dave specified the fact that we want to be able to handle both easy scenarios, where all the nodes are equal
```
layer2-config:
  interfaces: ["eno1", "eno2"]
address-pools:
  - name: default
    mode: layer2
    addresses:
      - 10.0.0.10 - 10.0.0.100
 ```

but also more complex scenarios where we want different configs based on the nodes:
```
address-pools:
  - name: default
    protocol: layer2
    addresses:
      - 10.0.0.5 - 10.0.100.10
    interfaces:
      - interfaces: ["eno1", "vlan6"]
        node-selectors:
          - match-labels:
              rack-name: foo
          - match-labels:
              rack-name: bar
      - interfaces: ["eno2"]
        node-selectors:
          - match-labels:
              rack-name: quux
```

Now, this being a property of L2Advertisement, and given the fact that L2Advertisement has a node selector, I _think_ we are covered in that regard. However, I'd document it explicitly (about the use case), and also I'd ensure it works with e2e tests.

Reply:
Thanks! @fedepaol @gclawes 
I reanalyzed my scenario in the morning.

![veth](https://user-images.githubusercontent.com/26404219/168544685-0b58a8cc-d9f3-4ac3-9d76-711d9422c95b.svg)

The above image describes my initial scenario. The fatal problem is the failure of network connectivity due to asymmetric routing.

For example:
```
 veth0 management network on 192.0.2.0/24
 veth1 data network on 198.51.100.0/24
```
Because MetalLB advertises all IPs from all interfaces, the management client may send a request to veth1, LB Service will reply from veth0 according to the route, which constitutes asymmetric routing so that the host will drop the packets sent by the management client.

My real need is for different subnet IPs to be advertised by the interface specified in node. Subnet IPs can be described by AddressPool. So the design goal is different AddressPools to be advertised by the interface specified in the specified node.
If you agree with that, I will revise my design according to the above goals. @fedepaol @gclawes